src/model/preprocess.py

```diff
@@ -0,0 +1,119 @@
+import regex as re
+
+# -----------------------  Implementation ---------------------
+def clean_doc(code: str, delim: str):
+    string_pre = r"(\S+\s*=\s*)"
+    between = rf"({delim})((?:(?!{delim})[\s\S])*)({delim})"
+
+    # this should not hit
+    holder = "!|<1multiline1>|!"
+
+    # initially save mutliline strings with flag value
+    code = re.sub(
+        string_pre + between,
+        rf"\1{holder}\3{holder}",
+        code,
+    )
+
+    # removing docs without caring about mutliline
+    code = re.sub(between, "", code)
+
+    # fix flags back to original
+    code = re.sub(re.escape(holder), delim, code)
+    return code
+
+
+# Im not sure this was the best approach for doc comments
+# Clean inline comments and block comments
+def clean_comments(code: str) -> str:
+    # remove comments with #
+    code = re.sub(r"#[^\n]*", "", code)
+
+    # remove docs with """
+    code = clean_doc(code, '"""')
+
+    # remove docs with '''
+    code = clean_doc(code, "'''")
+
+    # get rid of trailing
+    code = re.sub(r"\s*\n", "\n", code)
+    return code
+
+
+# for lambda deciding what files to include as training data
+def include(content: str) -> bool:
+    libraries = [
+        "numpy",
+        "pandas",
+        "matplotlib",
+        "sklearn",
+        "tensorflow",
+        "torch",
+        "scipy",
+    ]
+    for library in libraries:
+        a = "import " + library
+        b = "from " + library + " import"
+        if a in content or b in content:
+            return True
+    return False
+
+
+# Filters samples to return only code
+def keep_only_content(sample: dict) -> dict:
+    # TODO: Is there a way to remove other keys instead of forcing []?
+    for key in sample.keys():
+        if key == "content":
+            continue
+        sample[key] = []
```

Comment:
Could use `pop` to drop the keys we don't want. Also, should make the if condition check for `key != "content"` instead of `key == "content"` since `!=` is the common case
```suggestion
    for key in sample.keys():
        if key != "content":
            sample[key].pop()
```

Reply:
In general though I would question whether this function is really useful. Not sure if the work of iterating over the dict keys for each element is worth it given that this is a streaming dataset. Whether this is useful depends on `.map` implementation

Reply:
Implemented better version in https://github.com/MichiganDataScienceTeam/F24-mini-copilot/commit/48b7f4d66708990cf67b64f1afa218721a3dc67d